extra_data/tests/mockdata/agipd.py

```diff
@@ -0,0 +1,130 @@
+import numpy as np
+
+from .base import DeviceBase
+
+
+class AGIPDMDL(DeviceBase):
+    def __init__(
+        self, device_id,
+        rep_rate=True,
+        gain_setting=True,
+        integration_time=True
+    ):
+        super().__init__(device_id)
+        self.rep_rate = rep_rate
+        self.gain_setting = gain_setting
+        self.integration_time = integration_time
+
+        # A sample of some of the available keys.
+        self.control_keys = [
+            ('acquisitionTime', 'u2', ()),
+            ('bunchStructure.nPulses', 'u2', ()),
+            ('bunchStructure.firstPulse', 'u2', ()),
+            ('bunchStructure.methodIndex', 'u2', ()),
+            ('egsgat', 'u2', ()),
+            ('g1sgat', 'u2', ()),
+            ('g2sgat', 'u2', ()),
+            ('pcRowNr', 'u2', ()),
+            ('t0Delay', 'u8', ()),
+            ('ticolm', 'u2', ()),
+            ('vrfcds', 'u2', ()),
+        ]
+
+    def write_control(self, f):
+        super().write_control(f)
+
+        ctrl_grp = f'CONTROL/{self.device_id}'
+        run_grp = f'RUN/{self.device_id}'
+        for grp in [ctrl_grp, run_grp]:
+            length = self.ntrains if grp == ctrl_grp else 1
+
+            f[f"{grp}/setupr/value"] = np.array([32]*length)
+            f[f"{grp}/patternTypeIndex/value"] = np.array([4]*length)
+            f[f"{grp}/gainModeIndex/value"] = np.array([0]*length)
+
+            if self.integration_time:
+                f[f"{grp}/integrationTime/value"] = np.array([15]*length)
```

Comment:
For the ones that are optional, don't forget to check what the dtype is in real files and set it appropriately here, e.g. `np.full(length, 15, np.uint64)`  (I haven't checked, so uint64 is probably wrong).

Also, don't forget to create the `/timestamp` field manually for any keys that aren't in `control_keys`.

Reply:
Thinking about it, it's probably easier to conditionally add them to `control_keys` inside `__init__`, then they'll get timestamps created automatically, and you can just fill them in here.